resolwe/elastic/builder.py

```diff
@@ -0,0 +1,108 @@
+""".. Ignore pydocstyle D400.
+
+=====================
+Elastic Index Builder
+=====================
+
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from importlib import import_module
+import inspect
+
+from elasticsearch_dsl.connections import connections
+
+from django.apps import apps
+from django.conf import settings
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from .indices import BaseIndex
+
+
+__all__ = ('index_builder',)
+
+
+class ElasticSignal(object):
+    """Class for creating signals to update/create indexes.
+
+    To register the signal, add the following code::
+
+        from django.dispatch import receiver
+
+        signal = ElasticSignal(<my_signal>, <method_name>)
+        receiver(<signal_type>, [sender=<my_model>])(signal)
+
+    ``signal type`` can be i.e. ``django.db.models.signals.pre_save``.
+
+    """
+
+    def __init__(self, index, method_name):
+        """Initialize signal."""
+        self.index = index
+        self.method_name = method_name
+
+    def call(self, sender, instance, **kwargs):
+        """Process signal."""
+        method = getattr(self.index, self.method_name)
+        method(obj=instance)
+
+
+class IndexBuilder(object):
+    """Find indexes and register coresponding signals.
+
+    TODO
+
+    """
+
+    #: list of index builders
+    indexes = []
+
+    def __init__(self):
+        """Initialize index builder object."""
+        # Set dafault connection for ElasticSearch
+        # TODO: Reestablish connection if broken!
+        connections.create_connection(
+            hosts=['{}:{}'.format(settings.ELASTICSEARCH_HOST, settings.ELASTICSEARCH_PORT)])
+
+        self.discover_indexes()
+
+    def register_signals(self):
+        """Register signals for all indexes."""
+        for index in self.indexes:
+            if index.object_type:
+                post_save_signal = ElasticSignal(index, 'run')
+                receiver(post_save, sender=index.object_type)(post_save_signal)
+                pre_delete_signal = ElasticSignal(index, 'remove_object')
+                receiver(post_save, sender=index.object_type)(pre_delete_signal)
+
+    def discover_indexes(self):
+        """Save list of index builders into ``_index_builders``."""
+        for app_config in apps.get_app_configs():
+            indexes_path = '{}.elastic_indexes'.format(app_config.name)
+            try:
+                indexes_module = import_module(indexes_path)
+
+                for attr_name in dir(indexes_module):
+                    attr = getattr(indexes_module, attr_name)
+                    if inspect.isclass(attr) and issubclass(attr, BaseIndex) and attr is not BaseIndex:
+                        self.indexes.append(attr())
+            except ImportError:
+                pass  # no `elastic_indexes` in app
+
+    def trigger(self, obj=None):
```

Comment:
I'll name it `build`, because it triggers whole build process and `remove_object` only removes single object, so there should be some difference in the name.